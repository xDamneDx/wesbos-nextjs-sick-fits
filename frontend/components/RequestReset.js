import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

// Hooks:
import useForm from '../lib/useForm';

// Components:
import ErrorMessage from './ErrorMessage';

// Styled components:
import FormStyles from './styles/FormStyles';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    name: '',
    password: '',
  });

  const [signup, { data, loading, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  const handleSubmit = async (e) => {
    // Stop the form from submitting:
    e.preventDefault();

    // Send the email and password to the graphqlAPI:
    await signup().catch(console.error);
    resetForm();
  };

  return (
    <FormStyles method="POST" onSubmit={handleSubmit}>
      <h2>Request a Password Reset</h2>
      <ErrorMessage error={error} />
      <fieldset>
        {data?.sendUserPasswordResetLink === null && (
          <p>Success! Check your email for a link!</p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Email"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Request Reset!</button>
      </fieldset>
    </FormStyles>
  );
}
