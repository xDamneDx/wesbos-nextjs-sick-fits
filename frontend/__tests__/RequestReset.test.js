import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'email@gmail.com';
const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email, name: '', password: '' },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset/>', () => {
  it('Renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('calls the mutation when submitted', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    );
    // Type into the email box:
    userEvent.type(screen.getByPlaceholderText(/email/i), email);
    // Click submit:
    userEvent.click(screen.getByText(/Request Reset/));
    const success = await screen.findByText(/Success/i);
    expect(success).toBeInTheDocument();
  });
});
