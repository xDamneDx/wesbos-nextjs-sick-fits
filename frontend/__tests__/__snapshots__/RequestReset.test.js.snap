// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> Renders and matches snapshot 1`] = `
<div>
  <form
    class="FormStyles-sc-lfnj3o-0 dHqVho"
    method="POST"
  >
    <h2>
      Request a Password Reset
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Request Reset!
      </button>
    </fieldset>
  </form>
</div>
`;
